### ISSUE TITLE
Amalgamated build requires pkcs11.h

### ISSUE THREAD

**Issue by dmitry-xnor** [randombit/botan#976]
Amalgamated build requires pkcs11.h, which, in turn, requires a bunch of other stuff.

Repro (on a recent Intel Xeon):
```
git clone https://github.com/randombit/botan.git
cd botan
./configure.py --amalgamation --single-amalgamation-file
g++ -std=c++11 botan_all.cpp
```

Result:
```
In file included from botan_all.cpp:8:0:
botan_all.h:19302:20: fatal error: pkcs11.h: No such file or directory
compilation terminated.
```

Expected result:

`*.cpp` should compile and produce an object file.

**Comment by webmaster128** [comment]
Why don't you just call `make`? When you look into the Makefile, you see which include paths you need to add to your compiler call, if you want to call it directly

```
build/obj/lib/botan_all.o: botan_all.cpp
	$(CXX) $(LIB_FLAGS) -Ibuild/include -Ibuild/include/external -c botan_all.cpp -o $@
```

**Comment by dmitry-xnor** [comment]
Because I thought the whole point of amalgamated build was to create a self contained subset of files that can be compiled anywhere. `-Ibuild/include` would kind of defeat the purpose.

To quote the docs: "You can also configure Botan to be built using only a single source file; this is quite convenient if you plan to embed the library into another application."

**Comment by dmitry-xnor** [comment]
I mean, it's still not a "single file" under the best of circumstances, but even half a dozen would be a lot better for what I'm trying to do than a few hundred.

**Comment by webmaster128** [comment]
Right. I think `-Ibuild/include` is not necessary because you have everything from there combined in botan_all.h and botan_all_internal.h. pkcs11 is a special case where headers are used that are not in the library (I don't know why but there is a reason)

 ```
./configure.py --amalgamation --single-amalgamation-file --disable-modules=pkcs11
g++ -std=c++11 -c botan_all.cpp
```

works for me

**Comment by dmitry-xnor** [comment]
Maybe it'll work for us as well. Thanks for the suggestion. We need a small subset of ciphers, modes, and hashes. I'll try that and report back later today.

**Comment by dmitry-xnor** [comment]
Seems to work fine for us. We used the following flags:

`./configure.py --enable-modules=<list of modules> --amalgamation --minimized-build`

To reduce the object file size further. I'm closing the issue, feel free to reopen if pkcs11 dependency is unexpected.

**Comment by neusdan** [comment]
I think it's a bug. The amalgamation build should include the external headers into `botan_all.h`. What do you think @randombit ?

**Comment by randombit** [comment]
Bug, especially considering PKCS11 is enabled by default now. Including the external headers directly into botan_all.h does mean that the application cannot include a different version of the PKCS11 headers (say it includes the system version via p11-kit). I'm not sure how much of a problem this would be in practice for anyone, though.

Resolution is either include the headers into botan_all.h, or disable anything with external headers in the amalgamation build. Header inclusion is probably better.

@dmitry-xnor Re "I mean, it's still not a "single file" under the best of circumstances" you can avoid splitting out the C++ code by ISA extensions with `--single-amalgamation-file`, this requires a sufficiently recent GCC or Clang (one which supports the `target` function attribute), then there is just `botan_all.cpp`.
